File: client/pwa/manifest.js
// manifest.js — emit a valid, installable Web App Manifest.  PHASE P0.
//
// Pure JS (no build step). Produces an object that serializes to a spec-compliant
// `manifest.webmanifest`. The build step (Trilho C) writes the JSON to disk and
// injects <link rel="manifest"> into index.html; this module owns the *shape*.
//
// Installable-shaped means, per the PWA install criteria (Chromium/Lighthouse):
//   - name (or short_name)
//   - start_url
//   - display is one of "standalone" | "fullscreen" | "minimal-ui"
//   - icons[] includes at least a 192x192 and a 512x512 PNG, and at least one
//     icon with purpose containing "any" (maskable alone is not enough).
//
// See ../../docs/plan.md §7 P0 for the contract and tests/unit/test_pwa_manifest_*
// / tests/client/pwa-manifest.test.js for the goldens.

/**
 * @typedef {Object} ManifestIcon
 * @property {string} src       Icon URL, relative to the manifest scope.
 * @property {string} sizes     Space-separated WxH list, e.g. "192x192".
 * @property {string} type      MIME type, e.g. "image/png".
 * @property {string} [purpose] "any", "maskable", or "any maskable".
 */

/**
 * @typedef {Object} ManifestShortcut
 * @property {string} name                  Shortcut label.
 * @property {string} [short_name]          Shorter label for tight surfaces.
 * @property {string} [description]         Accessible description.
 * @property {string} url                   Target URL (within scope).
 * @property {ManifestIcon[]} [icons]       Optional shortcut icons.
 */

/**
 * @typedef {Object} ManifestOptions
 * @property {string} [name]               Full application name.
 * @property {string} [short_name]         Home-screen label (<= ~12 chars ideal).
 * @property {string} [description]        Human description.
 * @property {string} [start_url]          URL opened on launch. Default "/".
 * @property {string} [scope]              Navigation scope. Default "/".
 * @property {string} [display]            "standalone" | "fullscreen" | "minimal-ui" | "browser".
 * @property {string} [orientation]        e.g. "portrait", "any".
 * @property {string} [theme_color]        Toolbar color (CSS color).
 * @property {string} [background_color]   Splash background (CSS color).
 * @property {string} [lang]               BCP-47 language tag, e.g. "pt-BR".
 * @property {string} [dir]                "ltr" | "rtl" | "auto".
 * @property {string} [id]                 Stable app identity.
 * @property {ManifestIcon[]} [icons]      Icon set (defaults to the standard set).
 * @property {ManifestShortcut[]} [shortcuts]  P5 shortcuts (optional).
 * @property {Object} [share_target]       P5 share target (optional, passed through).
 * @property {Object[]} [file_handlers]    P5 file handlers (optional, passed through).
 * @property {string[]} [categories]       App store categories.
 */

/**
 * The default icon set: maskable 192/512 + a plain "any" 512 + apple-touch.
 * These are *references*; the build step is responsible for emitting the files.
 * @type {ManifestIcon[]}
 */
export const DEFAULT_ICONS = [
  { src: "/icons/icon-192.png", sizes: "192x192", type: "image/png", purpose: "any" },
  { src: "/icons/icon-512.png", sizes: "512x512", type: "image/png", purpose: "any" },
  { src: "/icons/maskable-192.png", sizes: "192x192", type: "image/png", purpose: "maskable" },
  { src: "/icons/maskable-512.png", sizes: "512x512", type: "image/png", purpose: "maskable" },
];

/** @type {Required<Pick<ManifestOptions, "name"|"short_name"|"description"|"start_url"|"scope"|"display"|"theme_color"|"background_color"|"lang"|"dir">>} */
const DEFAULTS = {
  name: "tempestweb app",
  short_name: "tempestweb",
  description: "A tempestweb application.",
  start_url: "/",
  scope: "/",
  display: "standalone",
  theme_color: "#111111",
  background_color: "#ffffff",
  lang: "pt-BR",
  dir: "auto",
};

/** Installable display modes — what buildManifest will accept without coercion. */
const INSTALLABLE_DISPLAYS = new Set(["standalone", "fullscreen", "minimal-ui"]);
/** All spec display modes (used only by the validator). */
const DISPLAY_VALUES = new Set(["standalone", "fullscreen", "minimal-ui", "browser"]);

/**
 * Build a Web App Manifest object from options, filling installable-shaped
 * defaults. Project config overrides every field; unknown extra keys
 * (shortcuts/share_target/file_handlers/categories/id/orientation) pass through.
 *
 * @param {ManifestOptions} [options] Project overrides.
 * @returns {Object} A manifest object ready to JSON.stringify.
 */
export function buildManifest(options = {}) {
  const opts = options || {};
  const icons = Array.isArray(opts.icons) && opts.icons.length > 0 ? opts.icons : DEFAULT_ICONS;

  /** @type {Object} */
  const manifest = {
    name: opts.name ?? DEFAULTS.name,
    short_name: opts.short_name ?? DEFAULTS.short_name,
    description: opts.description ?? DEFAULTS.description,
    start_url: opts.start_url ?? DEFAULTS.start_url,
    scope: opts.scope ?? DEFAULTS.scope,
    display: INSTALLABLE_DISPLAYS.has(opts.display) ? opts.display : DEFAULTS.display,
    theme_color: opts.theme_color ?? DEFAULTS.theme_color,
    background_color: opts.background_color ?? DEFAULTS.background_color,
    lang: opts.lang ?? DEFAULTS.lang,
    dir: opts.dir ?? DEFAULTS.dir,
    icons: icons.map((icon) => ({ ...icon })),
  };

  // Stable app identity defaults to the scope when not provided.
  manifest.id = opts.id ?? manifest.scope;

  if (opts.orientation) manifest.orientation = opts.orientation;
  if (Array.isArray(opts.categories) && opts.categories.length > 0) {
    manifest.categories = [...opts.categories];
  }

  // P5 extras pass through untouched when present.
  if (Array.isArray(opts.shortcuts) && opts.shortcuts.length > 0) {
    manifest.shortcuts = opts.shortcuts.map((s) => ({ ...s }));
  }
  if (opts.share_target) manifest.share_target = { ...opts.share_target };
  if (Array.isArray(opts.file_handlers) && opts.file_handlers.length > 0) {
    manifest.file_handlers = opts.file_handlers.map((h) => ({ ...h }));
  }

  return manifest;
}

/**
 * Serialize a manifest object to a JSON string for `manifest.webmanifest`.
 * @param {Object} manifest A manifest object (typically from buildManifest).
 * @param {number} [indent] Spaces of indentation (default 2; 0 for minified).
 * @returns {string} JSON text.
 */
export function emitManifest(manifest, indent = 2) {
  return JSON.stringify(manifest, null, indent);
}

/**
 * Validate the P5 manifest extras (shortcuts / share_target / file_handlers).
 *
 * These are progressive enhancements with uneven browser support, so this is a
 * shape check, not an install requirement. Returns the list of problems; empty
 * means the extras that ARE present are well-formed.
 *
 * @param {Object} manifest A manifest object.
 * @returns {string[]} Human-readable problems ([] when the extras are valid).
 */
export function validateExtras(manifest) {
  /** @type {string[]} */
  const errors = [];
  if (!manifest || typeof manifest !== "object") return ["manifest must be an object"];

  if (manifest.shortcuts !== undefined) {
    if (!Array.isArray(manifest.shortcuts)) {
      errors.push("shortcuts must be an array");
    } else {
      manifest.shortcuts.forEach((s, i) => {
        if (!s || typeof s.name !== "string") errors.push(`shortcuts[${i}].name is required`);
        if (!s || typeof s.url !== "string") errors.push(`shortcuts[${i}].url is required`);
      });
    }
  }

  if (manifest.share_target !== undefined) {
    const st = manifest.share_target;
    if (!st || typeof st !== "object") {
      errors.push("share_target must be an object");
    } else {
      if (typeof st.action !== "string") errors.push("share_target.action is required");
      const method = (st.method ?? "GET").toUpperCase();
      if (method === "POST" && typeof st.enctype !== "string") {
        errors.push("share_target with method POST requires an enctype");
      }
    }
  }

  if (manifest.file_handlers !== undefined) {
    if (!Array.isArray(manifest.file_handlers)) {
      errors.push("file_handlers must be an array");
    } else {
      manifest.file_handlers.forEach((h, i) => {
        if (!h || typeof h.action !== "string") errors.push(`file_handlers[${i}].action is required`);
        if (!h || typeof h.accept !== "object") errors.push(`file_handlers[${i}].accept is required`);
      });
    }
  }

  return errors;
}

/**
 * The default P5 extras a freshly-scaffolded app ships: a "Home" shortcut, a
 * POST share target, and a CSV file handler. References only — routes are wired
 * by the host app (share_target pairs with native.share, N2).
 * @type {{shortcuts: ManifestShortcut[], share_target: Object, file_handlers: Object[]}}
 */
export const DEFAULT_EXTRAS = {
  shortcuts: [
    { name: "Home", short_name: "Home", url: "/", description: "Open the app home" },
  ],
  share_target: {
    action: "/share-target",
    method: "POST",
    enctype: "multipart/form-data",
    params: { title: "title", text: "text", url: "url" },
  },
  file_handlers: [
    { action: "/open", accept: { "text/csv": [".csv"] } },
  ],
};

/**
 * Check whether a manifest object meets the baseline install criteria used by
 * Chromium/Lighthouse. Returns the list of problems; empty means installable.
 *
 * @param {Object} manifest A manifest object.
 * @returns {string[]} Human-readable validation errors ([] when installable).
 */
export function validateInstallable(manifest) {
  /** @type {string[]} */
  const errors = [];
  if (!manifest || typeof manifest !== "object") {
    return ["manifest must be an object"];
  }
  if (!manifest.name && !manifest.short_name) {
    errors.push("name or short_name is required");
  }
  if (!manifest.start_url) {
    errors.push("start_url is required");
  }
  if (!DISPLAY_VALUES.has(manifest.display) || manifest.display === "browser") {
    errors.push('display must be "standalone", "fullscreen" or "minimal-ui"');
  }

  const icons = Array.isArray(manifest.icons) ? manifest.icons : [];
  const hasSize = (size) =>
    icons.some((icon) => typeof icon.sizes === "string" && icon.sizes.split(/\s+/).includes(size));
  if (!hasSize("192x192")) errors.push("a 192x192 icon is required");
  if (!hasSize("512x512")) errors.push("a 512x512 icon is required");

  const hasAnyPurpose = icons.some((icon) => {
    const purpose = icon.purpose ?? "any";
    return purpose.split(/\s+/).includes("any");
  });
  if (!hasAnyPurpose) errors.push('at least one icon must have purpose "any"');

  return errors;
}

File: tempestweb/pwa/manifest.py
"""Web App Manifest emitter (P0/P5).

Builds an installable-shaped ``manifest.webmanifest`` at build time. The field
set and install criteria mirror the pure-JS emitter in
``client/pwa/manifest.js`` so the build-time output and any client-side checks
never drift.

Installable-shaped means (Chromium/Lighthouse baseline):
    - ``name`` or ``short_name``
    - ``start_url``
    - ``display`` in {"standalone", "fullscreen", "minimal-ui"}
    - ``icons`` includes a 192x192 and a 512x512 PNG, and at least one icon with
      purpose containing "any".
"""

from __future__ import annotations

import json
from dataclasses import dataclass, field
from pathlib import Path
from typing import Any

INSTALLABLE_DISPLAYS: frozenset[str] = frozenset(
    {"standalone", "fullscreen", "minimal-ui"}
)

#: Default icon set: plain "any" 192/512 + maskable 192/512. References only;
#: ``tempestweb.pwa.icons.emit_icons`` writes the actual files.
DEFAULT_ICONS: list[dict[str, str]] = [
    {
        "src": "/icons/icon-192.png",
        "sizes": "192x192",
        "type": "image/png",
        "purpose": "any",
    },
    {
        "src": "/icons/icon-512.png",
        "sizes": "512x512",
        "type": "image/png",
        "purpose": "any",
    },
    {
        "src": "/icons/maskable-192.png",
        "sizes": "192x192",
        "type": "image/png",
        "purpose": "maskable",
    },
    {
        "src": "/icons/maskable-512.png",
        "sizes": "512x512",
        "type": "image/png",
        "purpose": "maskable",
    },
]


@dataclass(slots=True)
class ManifestOptions:
    """Project overrides for the generated manifest.

    Every field defaults to an installable-shaped value; a project's
    ``tempestweb`` config overrides what it needs.

    Attributes:
        name: Full application name.
        short_name: Home-screen label.
        description: Human description.
        start_url: URL opened on launch.
        scope: Navigation scope.
        display: One of the installable display modes.
        theme_color: Toolbar color (CSS color).
        background_color: Splash background (CSS color).
        lang: BCP-47 language tag.
        dir: Text direction ("ltr" | "rtl" | "auto").
        orientation: Optional preferred orientation.
        app_id: Stable app identity; defaults to ``scope`` when None.
        icons: Icon set; defaults to ``DEFAULT_ICONS`` when empty.
        categories: App-store categories.
        shortcuts: P5 app shortcuts.
        share_target: P5 share target descriptor.
        file_handlers: P5 file handler descriptors.
    """

    name: str = "tempestweb app"
    short_name: str = "tempestweb"
    description: str = "A tempestweb application."
    start_url: str = "/"
    scope: str = "/"
    display: str = "standalone"
    theme_color: str = "#111111"
    background_color: str = "#ffffff"
    lang: str = "pt-BR"
    dir: str = "auto"
    orientation: str | None = None
    app_id: str | None = None
    icons: list[dict[str, str]] = field(default_factory=list)
    categories: list[str] = field(default_factory=list)
    shortcuts: list[dict[str, Any]] = field(default_factory=list)
    share_target: dict[str, Any] | None = None
    file_handlers: list[dict[str, Any]] = field(default_factory=list)


def build_manifest(options: ManifestOptions | None = None) -> dict[str, Any]:
    """Build a manifest object from options, filling installable defaults.

    Args:
        options: Project overrides. ``None`` uses every default.

    Returns:
        A JSON-able manifest object ready for ``emit_manifest``.
    """
    opts = options or ManifestOptions()
    display = opts.display if opts.display in INSTALLABLE_DISPLAYS else "standalone"
    icons = opts.icons if opts.icons else DEFAULT_ICONS

    manifest: dict[str, Any] = {
        "name": opts.name,
        "short_name": opts.short_name,
        "description": opts.description,
        "start_url": opts.start_url,
        "scope": opts.scope,
        "display": display,
        "theme_color": opts.theme_color,
        "background_color": opts.background_color,
        "lang": opts.lang,
        "dir": opts.dir,
        "icons": [dict(icon) for icon in icons],
    }

    # Stable app identity defaults to the scope.
    manifest["id"] = opts.app_id if opts.app_id is not None else manifest["scope"]

    if opts.orientation:
        manifest["orientation"] = opts.orientation
    if opts.categories:
        manifest["categories"] = list(opts.categories)

    # P5 extras pass through untouched when present.
    if opts.shortcuts:
        manifest["shortcuts"] = [dict(s) for s in opts.shortcuts]
    if opts.share_target:
        manifest["share_target"] = dict(opts.share_target)
    if opts.file_handlers:
        manifest["file_handlers"] = [dict(h) for h in opts.file_handlers]

    return manifest


def emit_manifest(manifest: dict[str, Any], indent: int = 2) -> str:
    """Serialize a manifest object to JSON text for ``manifest.webmanifest``.

    Args:
        manifest: A manifest object (typically from ``build_manifest``).
        indent: Spaces of indentation (0 for minified).

    Returns:
        The JSON string. ``ensure_ascii`` is False so accented names survive.
    """
    return json.dumps(manifest, indent=indent or None, ensure_ascii=False)


def write_manifest(
    dest: Path,
    options: ManifestOptions | None = None,
    indent: int = 2,
) -> Path:
    """Build and write ``manifest.webmanifest`` to ``dest``.

    Args:
        dest: Output file path (parent dirs are created).
        options: Project overrides.
        indent: JSON indentation.

    Returns:
        The path written.
    """
    manifest = build_manifest(options)
    dest.parent.mkdir(parents=True, exist_ok=True)
    dest.write_text(emit_manifest(manifest, indent=indent) + "\n", encoding="utf-8")
    return dest


def validate_installable(manifest: dict[str, Any]) -> list[str]:
    """Return install-criteria errors for a manifest ([] when installable).

    Mirrors ``validateInstallable`` in ``client/pwa/manifest.js``.

    Args:
        manifest: A parsed manifest object.

    Returns:
        Human-readable problems; empty when installable.
    """
    errors: list[str] = []
    if not isinstance(manifest, dict):
        return ["manifest must be an object"]
    if not manifest.get("name") and not manifest.get("short_name"):
        errors.append("name or short_name is required")
    if not manifest.get("start_url"):
        errors.append("start_url is required")
    if manifest.get("display") not in INSTALLABLE_DISPLAYS:
        errors.append('display must be "standalone", "fullscreen" or "minimal-ui"')

    icons = manifest.get("icons") or []

    def has_size(size: str) -> bool:
        return any(size in str(icon.get("sizes", "")).split() for icon in icons)

    if not has_size("192x192"):
        errors.append("a 192x192 icon is required")
    if not has_size("512x512"):
        errors.append("a 512x512 icon is required")
    if not any("any" in str(icon.get("purpose", "any")).split() for icon in icons):
        errors.append('at least one icon must have purpose "any"')
    return errors


def default_extras() -> dict[str, Any]:
    """Return the default P5 manifest extras a scaffolded app ships.

    A "Home" shortcut, a POST share target and a CSV file handler. References
    only — the host app wires the routes (share_target pairs with native.share).

    Returns:
        A dict with ``shortcuts``, ``share_target`` and ``file_handlers``.
    """
    return {
        "shortcuts": [
            {
                "name": "Home",
                "short_name": "Home",
                "url": "/",
                "description": "Open the app home",
            }
        ],
        "share_target": {
            "action": "/share-target",
            "method": "POST",
            "enctype": "multipart/form-data",
            "params": {"title": "title", "text": "text", "url": "url"},
        },
        "file_handlers": [{"action": "/open", "accept": {"text/csv": [".csv"]}}],
    }


def validate_extras(manifest: dict[str, Any]) -> list[str]:
    """Validate the P5 manifest extras (shortcuts/share_target/file_handlers).

    Progressive enhancements with uneven browser support, so this is a shape
    check, not an install requirement. Mirrors ``validateExtras`` in
    ``client/pwa/manifest.js``.

    Args:
        manifest: A manifest (or extras) object.

    Returns:
        Human-readable problems; empty when present extras are well-formed.
    """
    errors: list[str] = []
    if not isinstance(manifest, dict):
        return ["manifest must be an object"]

    shortcuts = manifest.get("shortcuts")
    if shortcuts is not None:
        if not isinstance(shortcuts, list):
            errors.append("shortcuts must be an array")
        else:
            for i, s in enumerate(shortcuts):
                if not isinstance(s, dict) or not isinstance(s.get("name"), str):
                    errors.append(f"shortcuts[{i}].name is required")
                if not isinstance(s, dict) or not isinstance(s.get("url"), str):
                    errors.append(f"shortcuts[{i}].url is required")

    share_target = manifest.get("share_target")
    if share_target is not None:
        if not isinstance(share_target, dict):
            errors.append("share_target must be an object")
        else:
            if not isinstance(share_target.get("action"), str):
                errors.append("share_target.action is required")
            method = str(share_target.get("method", "GET")).upper()
            if method == "POST" and not isinstance(share_target.get("enctype"), str):
                errors.append("share_target with method POST requires an enctype")

    file_handlers = manifest.get("file_handlers")
    if file_handlers is not None:
        if not isinstance(file_handlers, list):
            errors.append("file_handlers must be an array")
        else:
            for i, h in enumerate(file_handlers):
                if not isinstance(h, dict) or not isinstance(h.get("action"), str):
                    errors.append(f"file_handlers[{i}].action is required")
                if not isinstance(h, dict) or not isinstance(h.get("accept"), dict):
                    errors.append(f"file_handlers[{i}].accept is required")

    return errors
